pkg/deploy/controller/generictrigger/controller.go

```diff
@@ -0,0 +1,199 @@
+	if decoded == nil {
+		// The decoded deployment config will never be nil here but a sanity check
+		// never hurts.
+		return false, nil
+	}
+	ictCount, resolved, canTriggerByImageChange := 0, 0, false
+	var causes []deployapi.DeploymentCause
+
+	// IMAGE CHANGE TRIGGERS
+	for _, t := range config.Spec.Triggers {
+		if t.Type != deployapi.DeploymentTriggerOnImageChange {
+			continue
+		}
+		ictCount++
+
+		// If this is the initial deployment then we need to wait for the image change controller
+		// to resolve the image inside the pod template.
+		lastTriggered := t.ImageChangeParams.LastTriggeredImage
+		if len(lastTriggered) == 0 {
+			continue
+		}
+		resolved++
+
+		// Non-automatic triggers should not be able to trigger deployments.
+		if !t.ImageChangeParams.Automatic {
+			continue
+		}
+
+		// We need stronger checks in order to validate that this template
+		// change is an image change. Look at the deserialized config's
+		// triggers and compare with the present trigger.
+		if !triggeredByDifferentImage(*t.ImageChangeParams, *decoded) {
+			continue
+		}
+
+		canTriggerByImageChange = true
+		causes = append(causes, deployapi.DeploymentCause{
+			Type: deployapi.DeploymentTriggerOnImageChange,
+			ImageTrigger: &deployapi.DeploymentCauseImageTrigger{
+				From: kapi.ObjectReference{
+					Name:      t.ImageChangeParams.From.Name,
+					Namespace: t.ImageChangeParams.From.Namespace,
+					Kind:      "ImageStreamTag",
+				},
+			},
+		})
+	}
+
+	// We need to wait for all images to resolve before triggering a new deployment.
+	if ictCount != resolved {
+		return false, nil
+	}
+
+	// CONFIG CHANGE TRIGGERS
+	canTriggerByConfigChange := false
+	// Our deployment config has a config change trigger and no image change has triggered.
+	// If an image change had happened, it would be enough to start a new deployment without
+	// caring about the config change trigger.
+	if deployutil.HasChangeTrigger(config) && !canTriggerByImageChange {
+		// This is the initial deployment or the config has a template change. We need to
+		// kick a new deployment.
+		if config.Status.LatestVersion == 0 || !kapi.Semantic.DeepEqual(config.Spec.Template, decoded.Spec.Template) {
+			canTriggerByConfigChange = true
+			causes = []deployapi.DeploymentCause{{Type: deployapi.DeploymentTriggerOnConfigChange}}
+		}
+	}
+
+	return canTriggerByConfigChange || canTriggerByImageChange, causes
+}
+
+// triggeredByDifferentImage compares the provided image change parameters with those found in the
+// previous deployment config (the one we decoded from the annotations of its latest deployment)
+// and returns whether the two deployment configs have been triggered by a different image change.
+func triggeredByDifferentImage(ictParams deployapi.DeploymentTriggerImageChangeParams, previous deployapi.DeploymentConfig) bool {
+	for _, t := range previous.Spec.Triggers {
+		if t.Type != deployapi.DeploymentTriggerOnImageChange {
```

Comment:
It should be illegal:

1. To have two config change triggers of different automatic values
(validation), but we can collapse identical config change triggers
(strategy)
2. To have two image change triggers pointing to the same container with
different "from" fields (validation), but we should be able to resolve
overlapping triggers to a minimal set (strategy)

The validation would be higher priority than the normalization.


On Wed, Jun 15, 2016 at 5:40 AM, Michail Kargakis <notifications@github.com>
wrote:

> In pkg/deploy/controller/generictrigger/controller.go
> <https://github.com/openshift/origin/pull/9229#discussion_r67130582>:
>
> > +		// kick a new deployment.
> > +		if config.Status.LatestVersion == 0 || !kapi.Semantic.DeepEqual(config.Spec.Template, decoded.Spec.Template) {
> > +			canTriggerByConfigChange = true
> > +			causes = []deployapi.DeploymentCause{{Type: deployapi.DeploymentTriggerOnConfigChange}}
> > +		}
> > +	}
> > +
> > +	return canTriggerByConfigChange || canTriggerByImageChange, causes
> > +}
> > +
> > +// triggeredByDifferentImage compares the provided image change parameters with those found in the
> > +// previous deployment config (the one we decoded from the annotations of its latest deployment)
> > +// and returns whether the two deployment configs have been triggered by a different image change.
> > +func triggeredByDifferentImage(ictParams deployapi.DeploymentTriggerImageChangeParams, previous deployapi.DeploymentConfig) bool {
> > +	for _, t := range previous.Spec.Triggers {
> > +		if t.Type != deployapi.DeploymentTriggerOnImageChange {
>
> I will open a separate pull that handles normalizing duplicate ICTs. Where
> is the appropriate place? Validation? Defaulting? Strategy?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/origin/pull/9229/files/57e667b8e7492fef318cfd8c149e2c9b3bcf13ee..cdd0b308492cf4f6daad011e6ec01567027b5bcd#r67130582>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ABG_p-O8T9gpcCRVTtw1Fw3Y50_jFAShks5qL8ilgaJpZM4IxPRQ>
> .
>